torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py

```diff
@@ -141,6 +179,9 @@ def decompress(fut):
         q = fut.value()[0].div_(world_size)
         torch.matmul(p, q.t(), out=matrix)
```

Comment:
Could this be where the bug is? The `p` here is not from the scope of the `compute_q` callback, but from the parent scope. That tensor is still '*empty*'.

Reply:
The value of `p` in `decompress(fut)` is same as the value in `compute_q ` as well as the parent scope.

The last line `allreduce_p_fut.then(compute_q).then(decompress)` basically chains up the parent scope and the 2 callbacks.

Reply:
Ah, I see. My bad. I misunderstood the future-based all_reduce API. I thought it wouldn't modify its input and rather create a new tensor with the result of averaging.